### DISCUSSION TITLE
Chakra v3: how to extend the theme?

### DISCUSSION THREAD

**Discussion by jdddog** [chakra-ui/chakra-ui#8505] [Q&A] [answered]
Does anyone have any pointers about how to extend a theme in Chakra UI v3? 

I know it isn't actually released fully yet, however, I've been figuring out how to use it from the codebase and storybook stories, I'm just struggling a bit with custom theming (which seems to break when using the Next.js edge runtime).

The below example works in in a Next.js project with the app router and the default runtime. If the edge runtime is used the styling is broken and fails with a hydration? error for some reason. I'm not sure if I just haven't extended the theme properly or if it is something else. It works fine in the edge runtime if you use pass defaultSystem to ChakraProvider like in the next-app template.

![image](https://github.com/chakra-ui/chakra-ui/assets/5715104/1e9e035e-eb2f-4688-a285-774ebe9f101e)

```javascript
Warning: Prop `className` did not match. Server: "chakra-heading css-1dlnv1u" Client: "chakra-heading css-1likbrq"
    at h2
    at eval (
```

theme.ts (based on https://github.com/chakra-ui/chakra-ui/blob/main/packages/react/src/preset.ts):
```typescript
import {createSystem, defaultConfig, defineConfig, mergeConfigs} from "@chakra-ui/react"

const theme = defineConfig({
        theme: {
            recipes: {
                Button: {
                    variants: {
                        variant: {
                            custom: {
                                borderRadius: "full",
                                bg: "blue",
                                color: "white",
                                textTransform: "uppercase",
                            },
                        },
                    },
                },
            },
        }
    }
);

// Extends default theme
const config = mergeConfigs(defaultConfig, theme);
const system = createSystem(config);

export default system;
``` 

provider.tsx (based on the example next-app template): 
```typescript
"use client"

import {ChakraProvider} from "@chakra-ui/react"
import {ThemeProvider} from "next-themes"
import theme from "../theme/theme";

export default function RootLayout(props: { children: React.ReactNode }) {
    return (
        <ChakraProvider value={theme}>
            <ThemeProvider attribute="class" disableTransitionOnChange>
                {props.children}
            </ThemeProvider>
        </ChakraProvider>
    )
}
```

Thanks for your help.

**Comment by jdddog** [comment]
It seems that the issue is caused by the `mergeConfigs` function. It calls `lodash.mergeWith` which doesn't seem to work on the edge runtime. This means that the server render generates a different config than the client render, causing the mismatch. To fix it I replaced `mergeConfigs` with `merge` from the "deepmerge" package.

**Reply by segunadebayo** [reply]
Thanks for this feedback. Very valuable!

**Reply by jdddog** [reply]
You're welcome, thanks for creating a great project!

**Comment by Mounajidev** [comment]
Hello, we need to install a new pagacke to merge the configs ? i really dont get it, trying to start a new project with chakra v3.5, but when creating a theme seems not merging with the default base one, the mergeconfigs doesnt wor.. can anyone explain me more clearly how to achive that ? Thanks

**Reply by ghost** [reply]
Are you also using edge? the above merge solution works for me in firefox

**Reply by Mounajidev** [reply]
I am using Chrome, but i found my error was that i was using "defaultBaseConfig" instead of  " defaultConfig" . Thanks !

**Comment by segunadebayo** [comment, marked as answer]
Use `createSystem(defaultConfig, yours)`. Do not call `mergeConfigs` yourself.

The recipe key is `button`, not `Button`.

```ts
export const system = createSystem(defaultConfig, {
  theme: {
    recipes: {
      button: defineRecipe({
        variants: { variant: { custom: { bg: "blue" } } },
      }),
    },
  },
})
```
